sdk/src/types/block/mana/structure.rs

```diff
@@ -41,14 +45,53 @@ impl ManaStructure {
     }
 
     /// Returns the mana decay factor for the given epoch index.
-    pub fn decay_factor_at(&self, epoch_index: EpochIndex) -> Option<u32> {
-        self.decay_factors.get(*epoch_index as usize).copied()
+    pub fn decay_factor_at(&self, epoch_index: impl Into<EpochIndex>) -> Option<u32> {
+        self.decay_factors.get(*epoch_index.into() as usize).copied()
     }
 
     /// Returns the max mana that can exist with the mana bits defined.
     pub fn max_mana(&self) -> u64 {
         (1 << self.bits_count) - 1
     }
+
+    pub fn decay(&self, mana: u64, epoch_delta: u64) -> u64 {
+        if mana == 0 || epoch_delta == 0 {
+            return mana;
+        }
+
+        // split the value into two u64 variables to prevent overflowing
+        let mut mana_hi = upper_bits(mana);
+        let mut mana_lo = lower_bits(mana);
+
+        // we keep applying the lookup table factors as long as n epochs are left
+        let mut remaining_epochs = epoch_delta;
+
+        while remaining_epochs > 0 {
+            let epochs_to_decay = remaining_epochs.min(self.decay_factors().len() as u64);
+            remaining_epochs -= epochs_to_decay;
+
+            // Unwrap: Safe because the index is at most the length
+            let decay_factor = self.decay_factor_at(epochs_to_decay - 1).unwrap();
+
+            // apply the decay using fixed-point arithmetics.
+            (mana_hi, mana_lo) =
+                multiplication_and_shift(mana_hi, mana_lo, decay_factor, self.decay_factors_exponent());
+        }
```

Comment:
`epochs_to_decay` will either be the length of the decay factors lookup, or it will be `== remaining_epochs`, in which case subtracting it will make `remaining_epochs == 0`